src/partition/ReceivedPartition.cpp

```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+          {
+            vertexIsShared = true;
+            sharedVerticesSendMap[neighborRank.first].push_back(globalIDs[i]);
+            sharedVerticesList.push_back(globalIDs[i]);
+            sharedVerticesIndex.push_back(i);
+          }  
+        }
 
-    if (numberOfVertices != 0) {
-      PRECICE_DEBUG("Tag vertices, number of vertices " << numberOfVertices);
-      std::vector<int> tags(numberOfVertices, -1);
-      std::vector<int> globalIDs(numberOfVertices, -1);
-      bool             atInterface = false;
-      for (int i = 0; i < numberOfVertices; i++) {
-        globalIDs[i] = _mesh->vertices()[i].getGlobalIndex();
-        if (_mesh->vertices()[i].isTagged()) {
+        if(not vertexIsShared)
+        {
           tags[i]     = 1;
           atInterface = true;
-        } else {
-          tags[i] = 0;
+          counter++;
         }
       }
-      PRECICE_DEBUG("My tags: " << tags);
-      PRECICE_DEBUG("My global IDs: " << globalIDs);
-      PRECICE_DEBUG("Send tags and global IDs");
-      utils::MasterSlave::_communication->send(tags, 0);
-      utils::MasterSlave::_communication->send(globalIDs, 0);
-      utils::MasterSlave::_communication->send(atInterface, 0);
-
-      PRECICE_DEBUG("Receive owner information");
-      std::vector<int> ownerVec(numberOfVertices, -1);
-      utils::MasterSlave::_communication->receive(ownerVec, 0);
-      PRECICE_DEBUG("My owner information: " << ownerVec);
-      setOwnerInformation(ownerVec);
+      else
+      {
+          tags[i] = 0;
+      } 
     }
-  }
 
-  else if (utils::MasterSlave::isMaster()) {
-    // To temporary store which vertices already have an owner
-    std::vector<int> globalOwnerVec(_mesh->getGlobalNumberOfVertices(), 0);
-    // The same per rank
-    std::vector<std::vector<int>> slaveOwnerVecs(utils::MasterSlave::getSize());
-    // Global IDs per rank
-    std::vector<std::vector<int>> slaveGlobalIDs(utils::MasterSlave::getSize());
-    // Tag information per rank
-    std::vector<std::vector<int>> slaveTags(utils::MasterSlave::getSize());
-
-    // Fill master data
-    PRECICE_DEBUG("Tag master vertices");
-    bool masterAtInterface = false;
-    slaveOwnerVecs[0].resize(_mesh->vertices().size());
-    slaveGlobalIDs[0].resize(_mesh->vertices().size());
-    slaveTags[0].resize(_mesh->vertices().size());
-    for (size_t i = 0; i < _mesh->vertices().size(); i++) {
-      slaveGlobalIDs[0][i] = _mesh->vertices()[i].getGlobalIndex();
-      if (_mesh->vertices()[i].isTagged()) {
-        masterAtInterface = true;
-        slaveTags[0][i]   = 1;
-      } else {
-        slaveTags[0][i] = 0;
-      }
+    // #4: Exchange number of already owned vertices with the neighbors
+    std::map<int, int> neighborRanksVertexCount; // for load balancing
+    int receivedCounter = 0;
+    for (auto & neighborRank: localConnectedBBMap)
+    {
+      utils::MasterSlave::_communication->aSend(counter, neighborRank.first);
```

Comment:
I don't like the `aSend` here a lot. I guess it could still lead to a deadlock for very large cases depending on the MPI implementation.
Better could be:
* first `aReceive`
* then `send`
* last, check that all receives did happen

Reply:
I don't understand how this could result in a deadlock. Maybe I can check after aSend? 

I think what you suggest won't work, since here we first need to exchange size and if not zero communicate the vertex list. 
aReceive for size will always result in 0 size and no aReceive for vertex list will be called.

Reply:
I replaced aSend/receive with a aReceive/send. 

For the rest, let's continue the discussion below.